File: Golang/VehicleCounter/vehicleCounter.go
package main

import (
	"fmt"
	"image"
	"image/color"

	"gocv.io/x/gocv"
)

var (
	gray       = gocv.NewMat()
	result     = gocv.NewMat()
	before     = gocv.NewMat()
	after      = gocv.NewMat()
	isFirst    = true
	human      = 0
	motorcycle = 0
)

func main() {
	// declare window
	// originalWindow := gocv.NewWindow("original")

	red := color.RGBA{R: 255}

	// get video file
	videoPath := "../../data/vehicleCounter.avi"
	video, err := gocv.VideoCaptureFile(videoPath)
	if err != nil {
		fmt.Println("Cannot open the video")
	}
	defer video.Close()

	// create variable image
	img := gocv.NewMat()
	defer img.Close()

	// create template for erode and dilate
	erodeTemplate := gocv.GetStructuringElement(gocv.MorphRect, image.Point{X: 3, Y: 3})
	dilateTemplate := gocv.GetStructuringElement(gocv.MorphRect, image.Point{X: 35, Y: 35})

	if ok := video.Read(&img); !ok {
		fmt.Println("cannot read file")
	}
	// preparation to save result video
	saveFile := "../../result/result.avi"
	writer, err := gocv.VideoWriterFile(saveFile, "MJPG", 50, img.Cols()*2, img.Rows()*2, true)
	if err != nil {
		fmt.Printf("error opening video writer device: %v\n", err)
		return
	}
	defer writer.Close()

	// show the video frame by frame
	for {
		// to break terminal when video is end
		if ok := video.Read(&img); !ok || img.Empty() {
			break
		}

		// convert to gray image
		gocv.CvtColor(img, &gray, gocv.ColorBGRToGray)

		// save image before and now
		after = gray.Clone()
		if isFirst {
			before = after.Clone()
			result = after.Clone()
			isFirst = false
		}

		// subtract before and after
		gocv.AbsDiff(after, before, &result)

		// copy after to before
		before = after.Clone()

		gocv.Threshold(result, &result, 45, 255, gocv.ThresholdBinary)
		gocv.Erode(result, &result, erodeTemplate)
		gocv.Dilate(result, &result, dilateTemplate)

		moments := gocv.Moments(result, true)
		point := image.Point{
			X: int(moments["m10"] / moments["m00"]),
			Y: int(moments["m01"] / moments["m00"]),
		}

		// create dot (center of object)
		gocv.Circle(&img, point, 5, red, -1)

		// count blob and save it to variable
		contours := gocv.FindContours(result, gocv.RetrievalList, gocv.ChainApproxSimple)
		var pixels []int
		for idx, contour := range contours {
			if len(contour) > 75 {
				pixels = append(pixels, len(contour))
				gocv.DrawContours(&img, contours, idx, red, 2)
			}
		}

		for _, pixel := range pixels {
			// check the blob is already passed centerline or not yet
			if point.X > 293 && point.X < 310 {
				// check blob is motorcycle or human
				if pixel > 75 && pixel <= 110 {
					human++
				} else if pixel > 110 && pixel <= 150 {
					motorcycle++
				}
			}
		}

		text := fmt.Sprintf("human: %d, motorcycle: %d", human, motorcycle)

		// add centerline
		centerLine := img.Cols() / 2
		gocv.Line(&img, image.Point{X: centerLine}, image.Point{X: centerLine, Y: img.Rows()}, red, 2)

		// add text
		fmt.Println(text)
		gocv.PutText(&img, text, image.Point{X: 10, Y: img.Rows() - 50}, gocv.FontHersheySimplex, 1.1, red, 3)

		writer.Write(img)
		// originalWindow.IMShow(img)
		// originalWindow.WaitKey(50) // set 50 to normal video
	}
}

File: C++/VehicleCounter.cpp
#include "opencv2/imgproc.hpp"
#include "opencv2/imgcodecs.hpp"
#include "opencv2/highgui.hpp"
#include "iostream"

using namespace cv;
using namespace std;

int manusia = 0, motor = 0;

int blob(Mat img)
{
	vector<vector< Point >> contours;
	vector<Vec4i> hierarchy;
	findContours(img, contours, hierarchy, RETR_LIST, CHAIN_APPROX_SIMPLE);
	int blob = contours.size();
	int jml_blob=0,n=0,a=0;
	for (int n = 0; n < blob; n++)
	{
		if (contours[n].size() > 75)
		{
			jml_blob++;
			a = contours[n].size();
			//cout << "blob =" << jml_blob << "    ukuran blob =" << contours[n].size() << "\n" << endl;
		}
		
	}
	return a;
}

int main(int argc, char** argv)
{
	namedWindow("original", WINDOW_AUTOSIZE);
	namedWindow("result", WINDOW_AUTOSIZE);

	VideoCapture cap("../data/vehicleCounter.avi"); //cap.open(0);
	Mat frame; Mat gray; Mat it0; Mat it1; Mat hasil;
	Mat element2 = getStructuringElement(MORPH_RECT, Size(3, 3), Point(-1, -1));
	Mat element3 = getStructuringElement(MORPH_RECT, Size(35, 35), Point(-1, -1));
	int count = 0;
	Scalar rata2;
	for (;;)
	{
		cap >> frame;
		if (frame.empty()) break;
		cvtColor(frame, gray, COLOR_BGR2GRAY);
		
		//step 1
		it0 = gray.clone();
		if (count == 0)
		{
			it1 = it0.clone();
			count = 1;
			hasil = it0.clone();
		}
		
		//step 2
		absdiff(it0, it1, hasil);
		
		//step 3
		it1 = it0.clone();
		threshold(hasil, hasil, 45, 255, THRESH_BINARY);
		erode(hasil, hasil, element2);
		dilate(hasil, hasil, element3);
		Moments m = moments(hasil, true);
		Point p(m.m10 / m.m00, m.m01 / m.m00);
		int x = (int)(p).x;
		int y = (int)(p).y;
		circle(frame, p, 5, Scalar(0, 0, 255), -1);
		int pixel = blob(hasil);

		if (pixel > 75 && pixel < 110) {
			if (x > 295 && x < 310) {
				manusia++;
			}
		}
		if (pixel > 110 && pixel < 150) {
			if (x > 290 && x < 310) {
				motor++;
			}
		}

		printf("manusia:%d - motor:%d\n", manusia, motor);
		imshow("original", frame);
		imshow("result", hasil);
		char pencet = waitKey(10);
		if (pencet == 32) {
			while (pencet == 32) {
				if (waitKey(10) == 32) break; //pause
			}
		}
		else if (pencet == 27) return 0; //exit
	}
}